java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java

```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+          }
+        });
+        toReturn.put("args", stereotypeArguments);
+      }
+
+      if (name.equals("extensions")) {
+        List<String> extensionList = (List<String>) (capsOptions.get(("extensions")));
+
+        List<String> stereotypeExtensions =
+          (List<String>) (stereotypeOptions.getOrDefault(("extensions"), new ArrayList<>()));
+
+        extensionList.forEach(extension -> {
+          if (!stereotypeExtensions.contains(extension)) {
+            stereotypeExtensions.add(extension);
+          }
+        });
+
+        toReturn.put("extensions", stereotypeExtensions);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
+
+      if (!name.equals("binary") && !name.equals("extensions") && !name.equals("args")) {
+        toReturn.put(name, capsOptions.get(name));
+      }
     }
 
-    @SuppressWarnings("unchecked")
-    Map<String, Object> capsOptions = (Map<String, Object>) toReturn.get(options);
-    stereotypeOptions.forEach((key, value) -> {
-      if (!capsOptions.containsKey(key)) {
-        capsOptions.put(key, value);
+    return toReturn;
+  }
+
+  private Map<String, Object> mergeFirefoxOptions(Map<String, Object> stereotypeOptions,
+                                                  Map<String, Object> capsOptions) {
+    Map<String, Object> toReturn = new HashMap<>(stereotypeOptions);
+
+    for (String name : capsOptions.keySet()) {
+      if (name.equals("args")) {
+        List<String>
+          arguments =
+          (List<String>) (capsOptions.getOrDefault(("args"), new ArrayList<>()));
+        List<String> stereotypeArguments =
+          (List<String>) (stereotypeOptions.getOrDefault(("args"), new ArrayList<>()));
+        arguments.forEach(arg -> {
+          if (!stereotypeArguments.contains(arg)) {
+            stereotypeArguments.add(arg);
+          }
+        });
+        toReturn.put("args", stereotypeArguments);
+      }
+
+      if (name.equals("prefs")) {
+        Map<String, Object> prefs =
+          (Map<String, Object>) (capsOptions.getOrDefault(("prefs"), new HashMap<>()));
+
+        Map<String, Object> stereotypePrefs =
+          (Map<String, Object>) (stereotypeOptions.getOrDefault(("prefs"), new HashMap<>()));
+
+        stereotypePrefs.putAll(prefs);
+        toReturn.put("prefs", stereotypePrefs);
       }
-    });
 
-    return new ImmutableCapabilities(toReturn);
+      if (name.equals("profile")) {
+        String rawProfile =
+          (String) capsOptions.get("profile");
+
+        toReturn.put("profile", rawProfile);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
+
+      if (name.equals("log")) {
+        Map<String, Object> logLevelMap =
+          (Map<String, Object>) capsOptions.get("log");
+        toReturn.put("log", logLevelMap);
+      }
+    }
```

Comment:
Same as the `binary` case.

Reply:
This is log-level property. For this, we should consider the user preference here right?

Reply:
Ah OK. Log level should be fine.